### ISSUE TITLE
FAQ? How to install on Fedora.

### ISSUE THREAD

**Issue by kierun** [awesomeWM/awesome#1225]
**Output of `awesome --version`:**

`master` branch.

**How to reproduce the issue:**

Question really…

On Fedora, installing awesome from source can take two paths: either manual (`make install`) or via RPM (build from the source code, not pulled from `dnf`). Clearly, as I already have the system awesome installed, I would like to update that via the RPM. Does this do all the set up for gdm? If it just a drop-in replacement?

Although, I am getting this error:

```
Running make Makefile…
Running make package…
make[1]: *** No rule to make target 'package'.  Stop.
Makefile:43: recipe for target 'package' failed
make: *** [package] Error 2
```

**Comment by psychon** [comment]
> Does this do all the set up for gdm? If it just a drop-in replacement?

Dunno. If google served me well, then [here](http://pkgs.fedoraproject.org/cgit/rpms/awesome.git/tree/) is the source that Fedora uses to build awesome. You could clone that Git and try to replace the source code that is used with a newer version.

Looking at that git, differences are that the fallback-editor is changed to `vi` instead of `nano` and they apply some patch so that Lua is found. Doesn't seem to be much of a difference.

They seem to actually build awesome with `mkdir build ; pushd build ; %cmake -DAWESOME_DOC_PATH=SOMETHING1 -DSYSCONFDIR=SOMETHING2 .. ; popd ; make -C build VERBOSE=1 awesome`. Dunno what additional magic `%cmake` applies before calling the real CMake and dunno what the exact value for `SOMETHING1/2`.

@thmo You seem to be the Thomas Moschny mentioned in the Changelog for `awesome.spec`. Any insights?

For "`make package` does not work", I got this in the output from CMake:
```
-- rpmbuild not found.
-- The 'package' target will not build a RPM
-- dpkg not found.
-- The 'package' target will not build a DEB
```
So you need to install `rpmbuild` to get an .rpm.

**Comment by kierun** [comment]
> For "make package does not work", I got this in the output from CMake: […]

Yup, just found that in the output. I feel stupid now.

**Comment by kierun** [comment]
I just generated `awesome-3.5.2~git2180-g5aa4a16-Linux.rpm` form the master checkout. However, the currently installed version of awesome is v3.5.9 (Mighty Ravendark). Have I missed something? Should I have been building from a branch (which one?) to get HiDPI support?

**Comment by psychon** [comment]
Use `-DAWESOME_VERSION=9999` or something like that to explicitly set a version.
The version number is auto-generated via `git describe` and version 3.5.2 is the last common release between the 3.5 branch and the master branch (that's when they diverged). The exact version number means: You are on awesome 3.5.2 plus 2180 commits. The Git commit ID of what you built is 5aa4a16.

**Comment by kierun** [comment]
AH, of course. I forgot about that `git` tool since I have been abusing `bumpversion` for most of my work.

Thanks!

**Comment by thmo** [comment]
Yes, I am the maintainer for awesome in Fedora. [Here](https://kojipkgs.fedoraproject.org//packages/awesome/3.5.9/1.fc24/data/logs/x86_64/build.log) is the build log of the package in F24, and there you can also see how it is built (i.e. the `%cmake` expansion):

```
+ CFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic'
+ export CFLAGS
+ CXXFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic'
+ export CXXFLAGS
[...]
+ LDFLAGS='-Wl,-z,relro -specs=/usr/lib/rpm/redhat/redhat-hardened-ld'
+ export LDFLAGS
+ /usr/bin/cmake -DCMAKE_C_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_Fortran_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -DCMAKE_INSTALL_PREFIX:PATH=/usr -DINCLUDE_INSTALL_DIR:PATH=/usr/include -DLIB_INSTALL_DIR:PATH=/usr/lib64 -DSYSCONF_INSTALL_DIR:PATH=/etc -DSHARE_INSTALL_PREFIX:PATH=/usr/share -DLIB_SUFFIX=64 -DBUILD_SHARED_LIBS:BOOL=ON -DAWESOME_DOC_PATH=/usr/share/doc/awesome -DSYSCONFDIR=/etc ..
```

Now, I might be missing some context, but is there something wrong with the Fedora package? 3.5.9 is the latest version, isn't it? So we are not lagging behind.

**Comment by psychon** [comment]
@thmo Nope, you are not lagging behind. We are lagging behind in releasing. The last common commit on the master and 3.5 branch is c855b1babb0b2e8c171e16ea6f40da668d98a372 which is about 2.5 years old. People are starting to notice that master has some interesting, unreleased stuff and want to play with it. (And then they notice that the build will identify itself as 3.5.2-something-weird and then they wonder what is going on...).

**Comment by kierun** [comment]
> Now, I might be missing some context, but is there something wrong with the Fedora package? 3.5.9 > is the latest version, isn't it? So we are not lagging behind.

Not as such… I have a new shiny 4k monitor and the fonts on the `wibar` look tiny. I upped them in the `theme.lua` and that worked but not for the `vicious` widgets.

Maybe I am doing it wrong and am shaving Yaks…

On a side note, @thmo, many thanks indeed for maintaining the package. It is very much appreciated.

**Comment by kierun** [comment]
What I am driving at with all these is this: *Will the `make package` RPM (with appropriately increased version) install cleanly on Fedora 24?*

If not, what do I need to do to make it install cleanly?

From there onwards, I should be able to use [issue 760](https://github.com/awesomeWM/awesome/issues/760) to get things working.

**Comment by actionless** [comment]
i think you can either explicitly set version to a higher value as psychon suggested or you can just force install the package and ignore its update in package manager config

**Comment by kierun** [comment]
Any Fedora users that can confirm this works? 

Or at least way to test it without trashing your system?

**Comment by psychon** [comment]
> Any Fedora users that can confirm this works?

Commands (in awesome source dir):
```
rm -r .build*
make CMAKE_ARGS="-DCMAKE_INSTALL_PREFIX:PATH=/usr -DAWESOME_DOC_PATH=/usr/share/doc/awesome -DSYSCONFDIR=/etc " all package
sudo rpm -U --oldpackage build/awesome*rpm
```
Result:
```
	file /etc/xdg from install of awesome-3.5.2~git2189_gd59f943-1.x86_64 conflicts with file from package filesystem-3.2-37.fc24.x86_64
	file /usr/share/xsessions from install of awesome-3.5.2~git2189_gd59f943-1.x86_64 conflicts with file from package filesystem-3.2-37.fc24.x86_64
```
Sigh. Google finds https://ask.fedoraproject.org/en/question/37185/file-conflict-for-installing-a-package-with-filesystem/, so I do that (remove the entries for `/etc/xdg/` and `/usr/share/xsessions/`) and I get a new RPM. This one can be installed.
Afterwards I did
```
Xephyr :1 &
DISPLAY=:1 awesome -c /etc/xdg/awesome/rc.lua
```
and awesome started. So I guess this means everything works, right?

> Or at least way to test it without trashing your system?

Just test it. In case something breaks, you can always just go back to the version provided by Fedora (Since I don't know exactly how, I'd propose: Uninstall awesome, re-install awesome; Of course it's nice if you have e.g. XFCE or twm installed so that even when awesome breaks you still get a graphical session)
(In my case: I could just do `dnf upgrade` to get back to the Fedora version, since mine had a lower version number)
This is the nice thing about package managers which one does not get from `sudo make install`.

**Comment by kierun** [comment]
Tried it… It did not go well.

First, the `-DAWESOME_VERSION=9999` did nothing. The version build was still the old one. Fine, I do not mind that much.

Second, I could not get it working under `Xephyr`: awful was missing some missing things. Log attached although that is a mix of the old and new, so not ideal. Any idea how I could run a better check?
[log.zip](https://github.com/awesomeWM/awesome/files/621429/log.zip)

**Comment by psychon** [comment]
First error in the log is:
```
error: /usr/share/awesome/lib/awful/tag.lua:635: attempt to call a nil value (method 'add_signal')
```
Uhm, this is tag.lua from awesome 3.5, not from git/master. The line number also matches: https://github.com/awesomeWM/awesome/blob/3.5/lib/awful/tag.lua.in#L635
Shouldn't this be the file from master instead?

**Comment by kierun** [comment]
Yes, it should be. I did get a similar error but muppet that I am the `~/.zsession-error` file I had it in got over written.  I could re-install the RPM, run Zephir, watch it crash, and then have to re-install awesome from the normal repo. However, if there was a simpler way to test it, that would be grand!

**Comment by psychon** [comment]
Could you try removing/uninstalling awesome, making sure that really is no `/usr/share/awesome/lib/awful/tag.lua` file any more and then trying to install the selfmade rpm? That should make really, really, *really* sure that no old file is laying around.
(Uninstalling awesome while it is running should "just work" and have not much bad effects (some icons might go missing later on))

**Comment by kierun** [comment]
AHA! Progress. There must have been some old files in `/usr/share/awesome/awful`. Now, awesome does start. However:

```
2016-11-30 10:44:23 W: awesome: screen_scan_randr_crtcs:398: Ignoring RandR, only a compatibility layer is present.
2016-11-30 10:44:23 /usr/share/awesome/lib/beautiful/xresources.lua:54: W: Failed to get xrdb value 'background'.
stack traceback:
	/usr/share/awesome/lib/awful/hotkeys_popup/widget.lua:107: in function 'awful.hotkeys_popup.widget.new'
	/usr/share/awesome/lib/awful/hotkeys_popup/widget.lua:460: in upvalue 'get_default_widget'
	/usr/share/awesome/lib/awful/hotkeys_popup/widget.lua:477: in function 'awful.hotkeys_popup.widget.add_hotkeys'
	/usr/share/awesome/lib/awful/hotkeys_popup/keys/vim.lua:171: in main chunk
	[C]: in function 'require'
	/usr/share/awesome/lib/awful/hotkeys_popup/keys/init.lua:11: in main chunk
	[C]: in function 'require'
	/usr/share/awesome/lib/awful/hotkeys_popup/init.lua:12: in main chunk
	[C]: in function 'require'
	/etc/xdg/awesome/rc.lua:12: in main chunk
2016-11-30 10:44:23 W: beautiful: can't get colorscheme from xrdb (using fallback).
```

Also, does the `RPM` modify `gdm` so I can use Awesome as a window manager session from the login screen?

**Comment by psychon** [comment]
Just ignore those warnings (first one: You are running awesome in Xephyr; other two: There is no color scheme defined via `xrdb`).
> Also, does the RPM modify gdm so I can use Awesome as a window manager session from the login screen?

I didn't check that. However, I *think* it should do so. At least Fedora's official RPM doesn't add any special magic here. You can check if `/usr/share/xsessions/awesome.desktop` exists to make sure.
My content here with the official RPM is:
```
[Desktop Entry]
Encoding=UTF-8
Name=awesome
Comment=Highly configurable framework window manager
TryExec=awesome
Exec=awesome
Type=Application
```

**Comment by kierun** [comment]
… And this seemed to have worked!  I am impressed.

Thank you a lot for all the help. It is much appreciated. I will write up what I have done so that it can be added to documentation or whatnot.

Again, **thank you!!!**

**Comment by Elv13** [comment]
Closing this because 4.0 has been released and the 3.5.2-2000 < 3.5.9 issue is fixed
